### ISSUE TITLE
Returning 404s from components

### ISSUE THREAD

**Issue by kevinsimper** [reactjs/react-router#3210]
tldr; I did search and found this closed issue, where server-rendering was not taking into account https://github.com/reactjs/react-router/issues/357 

https://github.com/reactjs/react-router/issues/2286 is on the way, but that idea is more about regex and not about being able to return 404 after a database call.

I am using react-router server-side to render my app. I have a route file that I share between front-end and back-end.

```
export default (
  <Route path='/' component={Main}>
    <IndexRoute component={HelloWorld}/>
    <Route path='countries' component={Countries}/>
    <Route path='countries/:country' component={Country}/>
  </Route>
)
```

and I use this snippet from the react-router docs

```
export default (req, res, cb) => {
  match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
    if (error) {
      res.status(500).send(error.message)
    } else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search)
    } else if (renderProps) {
      console.log(renderProps)
      cb(renderToString(<RouterContext {...renderProps} />))
    } else {
      res.status(404).send('Not found')
    }
  })
}
```

If i go to `countries/:country` on url `countries/unitedkingdom` all is fine!

If I go to `countries/DOESNOT"#!€EXIST` the server should return 404, because else you can end up having a lot of empty pages indexed in google.

However it does not seem it is possible to return a 404 once RouterContext is started to render.

## Solution 1
Handle 404 before even beginning to render.

Upside: Simple
Drawback:
Needs to duplicate my routes, to both be in react-router format and in my express.js server.

## Solution 2
Be able to return 404 from within react component.

Upside: Keeps logic in react components
Drawback: Not possible right now

**Comment by timdorr** [comment]
This will be handled by something like #3098 (perhaps not that exact PR, though). We need some API to fail a route match (producing a 404) via a callback. 

For now, you can simply do a `this.context.router.replace` to a 404 route on your `componentWillMount. It's not a direct 404 (the return code will initially be a 301), but it works to the same effect in a pinch.

**Comment by taion** [comment]
Also a dupe of https://github.com/reactjs/react-router/issues/3181.

It's probably going to be #3098 – as in it can't be within the context of a component. Since the PR came first, let's track progress on the PR and open a new issue if needed.

**Comment by timdorr** [comment]
See #3212

**Comment by taion** [comment]
That actually orthogonal I think – for #3212 I just want to simplify getting a 404 status from the server.

**Comment by ryanflorence** [comment]
https://github.com/ryanflorence/react-project/blob/master/create-react-project/blueprint/modules/routes.js

**Comment by pavelkornev** [comment]
@kevinsimper have you found any adequate solution?

**Comment by kevinsimper** [comment]
@pavelkornev I reimplemented the routes in express, it was the easiest, because it did not complicate my components further :) I also use this to populate my redux store.

https://github.com/kevinsimper/cyclesdk/blob/master/app/server.js#L30-L46
